lib/Common/Memory/PageAllocator.cpp

```diff
@@ -749,70 +749,113 @@ PageAllocatorBase<T>::AddPageSegment(DListBase<PageSegmentBase<T>>& segmentList)
 }
 
 template<typename T>
-template <bool notPageAligned>
 char *
-PageAllocatorBase<T>::TryAllocFreePages(uint pageCount, PageSegmentBase<T> ** pageSegment, PageHeapMode pageHeapFlags)
+PageAllocatorBase<T>::TryAllocFromZeroPagesList(uint pageCount, PageSegmentBase<T> ** pageSegment, SLIST_HEADER& zeroPagesList)
 {
-    Assert(!HasMultiThreadAccess());
-    if (this->freePageCount < pageCount)
+    char * pages = nullptr;
+    FreePageEntry * freePage = (FreePageEntry *)::InterlockedPopEntrySList(&zeroPagesList);
+    FreePageEntry* localList = nullptr;
+    while (freePage != nullptr) {
+#if DBG
+        UpdateMinimum(this->debugMinFreePageCount, this->freePageCount);
+#endif
+
+        if (freePage->pageCount == pageCount)
+        {
+            *pageSegment = freePage->segment;
+            pages = (char *)freePage;
+            memset(pages, 0, sizeof(FreePageEntry));
+            this->FillAllocPages(pages, pageCount);
+            break;
+        }
+        else if (freePage->pageCount > pageCount)
+        {
+            *pageSegment = freePage->segment;
+            freePage->pageCount -= pageCount;
+            pages = (char *)freePage + freePage->pageCount * AutoSystemInfo::PageSize;
+            this->FillAllocPages(pages, pageCount);
+
+            ::InterlockedPushEntrySList(&backgroundPageQueue->freePageList, freePage);
+            break;
+        }
+        else
+        {
+            freePage->Next = localList;
+            localList = (FreePageEntry*)freePage;
+        }
+
+        freePage = (FreePageEntry *)::InterlockedPopEntrySList(&zeroPagesList);
+    }
+
+    while (localList != nullptr)
     {
-        return nullptr;
+        FreePageEntry* next = (FreePageEntry*)localList->Next;
+        ::InterlockedPushEntrySList(&zeroPagesList, localList);
```

Comment:
Same comment as before. You can't put it back to the pendingZeroPageList as it may be missed by the background GC thread's BackgroundZeroQueuedPages calls.

Alternatively, you will have to add a "last chance" ZeroQueuedPages on the main thread before we to ConcurrentTransferSwept.